tools/depcheck/pkg/cmd/analyze.go

```diff
@@ -0,0 +1,273 @@
+	return nil
+}
+
+func (o *AnalyzeOptions) Validate() error {
+	if len(o.Entrypoints) == 0 {
+		return errors.New("at least one root package must be provided")
+	}
+	if len(o.Dependencies) == 0 {
+		return errors.New("at least one dependency import path must be provided")
+	}
+
+	return nil
+}
+
+func (o *AnalyzeOptions) Run() error {
+	pkgs, err := graph.CalculatePackages(
+		ensureEntrypointEllipsis(
+			ensureVendorEntrypoint(o.Entrypoints, o.RepoRoot),
+		),
+	)
+	if err != nil {
+		return err
+	}
+
+	g, err := graph.BuildGraph(pkgs, o.GraphOptions.Entrypoints, o.Excludes)
+	if err != nil {
+		return err
+	}
+
+	if len(o.Filters) > 0 {
+		g, err = graph.FilterPackages(g, o.Filters)
+		if err != nil {
+			return err
+		}
+	}
+
+	// calculate possible roots
+	for _, n := range g.Nodes() {
+		node, ok := n.(*graph.Node)
+		if !ok {
+			continue
+		}
+		if !isOriginPackage(node, o.GraphOptions.RepoRoot) {
+			continue
+		}
+		if len(g.To(n)) > 0 || len(g.From(n)) == 0 {
+			continue
+		}
+
+		o.possibleRoots = append(o.possibleRoots, node.UniqueName)
+	}
+
+	// prune unreachable nodes from graph
+	g.PruneOrphans(o.possibleRoots)
+	return o.analyzeGraph(g)
+}
+
+// analyzeGraph receives a MutableDirectedGraph and outputs
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *AnalyzeOptions) analyzeGraph(g *graph.MutableDirectedGraph) error {
+	yoursToRemove := []*graph.Node{}
+	for _, dep := range o.Dependencies {
+		n, exists := g.NodeByName(dep)
+		if !exists {
+			return fmt.Errorf("unable to find dependency with import path %q", dep)
+		}
+		node, ok := n.(*graph.Node)
+		if !ok {
+			return fmt.Errorf("expected node to analyze to be of type *graph.Node. Got: %v", n)
+		}
+
+		yoursToRemove = append(yoursToRemove, node)
+	}
+
+	fmt.Printf("Analyzing a total of %v packages\n", len(g.Nodes()))
+	fmt.Println()
+
+	// calculate packages unique to the dependency that we are analyzing.
+	yours := graph.FindExclusiveDependencies(g, yoursToRemove, o.possibleRoots)
+	fmt.Printf("\"Yours\": %v packages exclusive to %q\n", len(yours), yoursToRemove)
+	for _, n := range yours {
+		fmt.Printf("    - %s\n", n.LabelName)
+	}
+	fmt.Println()
+
+	// calculate root repo packages
+	repoNodes := []*graph.Node{}
+	for _, n := range g.Nodes() {
+		node, ok := n.(*graph.Node)
+		if !ok {
+			return fmt.Errorf("expected node to analyze to be of type *graph.Node. Got: %v", n)
+		}
+		if !isOriginPackage(node, o.RepoRoot) {
+			continue
+		}
+
+		repoNodes = append(repoNodes, node)
+	}
+
+	// TODO: populate mine with repoNodes, immediate vendor deps of repoNodes,
+	// and ensure no nodes overlap with those in the "yours" set.
+	mine := []*graph.Node{}
```

Comment:
@deads2k I plan to address https://github.com/openshift/origin/pull/18466#issuecomment-368015835 here